cppcache/integration/test/ThinClientHARegionCacheListenerTest.cpp

```diff
@@ -0,0 +1,426 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <framework/Cluster.h>
+#include <framework/Framework.h>
+#include <framework/Gfsh.h>
+#include <gmock/gmock.h>
+
+#include <future>
+#include <iostream>
+#include <random>
+#include <thread>
+
+#include <boost/log/core.hpp>
+
+#include <gtest/gtest.h>
+
+#include <geode/AttributesMutator.hpp>
+#include <geode/Cache.hpp>
+#include <geode/CacheListener.hpp>
+#include <geode/EntryEvent.hpp>
+#include <geode/PoolManager.hpp>
+#include <geode/RegionEvent.hpp>
+#include <geode/RegionFactory.hpp>
+#include <geode/RegionShortcut.hpp>
+
+#include "mock/CacheListenerMock.hpp"
+
+namespace {
+// Until we get C++14 support...
+template <typename T, typename... Args>
+::std::unique_ptr<T> make_unique(Args &&... args) {
+  return ::std::unique_ptr<T>(new T(::std::forward<Args>(args)...));
+}
+
+// A simple comparator for cachables wrapped in shared pointers.
+MATCHER_P(CashableEq, value, "") {
+  return arg->toString() == value->toString();
+}
+
+class ThinClientHARegionCacheListenerTest : public ::testing::Test {
```

Comment:
I think we can drop the "ThinClient" prefix from all these tests in the future. All tests are for the "thin client" The original C++ was intended to be a peer "heavy client".

Reply:
Also will do.